src-ts/proposalMonitorCli.ts

```diff
@@ -1,70 +1,270 @@
-    options.governorAddress,
-    govChainSigner.provider!,
-    options.pollingIntervalMinutes * 60 * 1000,
-    options.blockLag,
-    options.startBlock,
-    pipelineFactory
-  );
 
-  proposalMonitor.on(GPMEventName.TRACKER_STARTED, (e: GPMStatusEvent) => {
-    console.log(`Tracker started:  Gov:${e.governorAddress}, Prop:${e.proposalId}`);
+const attachConsole = (proposalMonitor: ProposalMonitor, monitorName: string) => {
+  proposalMonitor.on(TrackerEventName.TRACKER_STARTED, (e: GPMEvent) => {
+    console.log(`${monitorName} tracker started:  Gov:${e.originAddress}, Stage:${e.identifier}`);
   });
 
-  proposalMonitor.on(GPMEventName.TRACKER_ENDED, (e: GPMStatusEvent) => {
-    console.log(`Tracker ended:  Gov:${e.governorAddress}, Prop:${e.proposalId}`);
+  proposalMonitor.on(TrackerEventName.TRACKER_ENDED, (e: GPMEvent) => {
+    console.log(`${monitorName} tracker ended:  Gov:${e.originAddress}, Prop:${e.identifier}`);
   });
 
-  proposalMonitor.on(GPMEventName.TRACKER_ERRORED, (e: GPMStatusEvent) => {
-    console.log(`Tracker errored:  Gov:${e.governorAddress}, Prop:${e.proposalId} Error:`, e);
+  proposalMonitor.on(TrackerEventName.TRACKER_ERRORED, (e: GPMErrorEvent) => {
+    console.log(
+      `${monitorName} tracker errored:  Gov:${e.originAddress}, Prop:${e.identifier} Stage:${
+        e.stage
+      } Status:${ProposalStageStatus[e.status]} Error: ${e.error}`
+    );
   });
-  proposalMonitor.on(GPMEventName.TRACKER_STATUS, (e: GPMStatusEvent) => {
+  proposalMonitor.on(TrackerEventName.TRACKER_STATUS, (e: GPMEvent) => {
     console.log(
-      `Gov:${e.governorAddress}, Prop:${e.proposalId}, Stage:${e.stage}, Status:${
-        ProposalStageStatus[e.status]
-      }`
+      `${monitorName} status update:  Gov:${e.originAddress}, Prop:${e.identifier}  Stage:${
+        e.stage
+      } Status:${ProposalStageStatus[e.status]}`
     );
   });
-  await proposalMonitor.start();
 };
 
-main().then(() => console.log("Done."));
\ No newline at end of file
+const startMonitor = async (
+  monitorName: string,
+  monitor: ProposalMonitor,
+  jsonLogger?: JsonLogger,
+  proposalId?: string
+) => {
+  attachConsole(monitor, monitorName);
+
+  if (jsonLogger) {
+    jsonLogger.subscribeToMonitor(monitor);
+  }
+  if (proposalId) {
+    const receipts = await monitor.getOriginReceipts(options.startBlock, "latest", proposalId);
+    if (receipts.length !== 1) {
+      throw new Error(`Proposal not found: ${proposalId}`);
+    }
+    await monitor.monitorSingleProposal(receipts[0]);
+  } else {
+    await monitor.start();
+  }
+};
+
+interface PipelineStage {
+  name: string;
+  identifier: string;
+  status: string;
+  explorerLink?: string;
+  proposalLink?: string;
+  children: PipelineStage[];
+}
+
+interface GovernorStatus {
+  [governorAddress: string]: PipelineStage[];
+}
+
+class JsonLogger {
+  constructor(public readonly fileLocation: string, public flushingIntervalMs: number) {}
+
+  private data: GovernorStatus = {};
```

Comment:
Having this object public & easily accessible in memory might be useful for e.g., the tally ui

Reply:
This GovernorStatus is only used by the logger. I think if Tally want to use this CLI in memory they want to refactor main() generally - and they would probably want to refactor out the logging part from the json updating part.